protocol/smart-contract-versioning.md

```diff
@@ -0,0 +1,86 @@
+- **Individual Contracts**: The current versioning scheme of individual contracts is also not a true semver, but in a different way. We do bump major and minor versions appropriately, but we don't distinguish when a contract is in beta, release candidate, or feature status. This makes it difficult to determine the stability and compatibility of a given contract version.
+
+These have resulted in various issues, such as chains deploying in-development versions of contracts.
+
+Unlike typical software development, OP Stack chains cannot rely on "compatible semvers" to compose the chain of various contracts.
+Instead, a given monorepo contracts release must be defined by an exact set of contract versions, not a range of semver-compatible versions.
+
+# Alternatives Considered
+
+1. Keep the current versioning scheme: This alternative would maintain the status quo but would not address the issues of clarity, consistency, and alignment with semver practices.
+
+2. Use semver for contracts but not for monorepo releases: This alternative would improve the versioning of individual contracts but would not provide a clear mapping between monorepo releases and contract versions.
+
+# Proposed Solution
+
+The proposed solution is to more closely adopt a true semver format for both individual contracts and monorepo releases.
+
+We preserve the existing rules for when to bump major, minor, and patch versions in individual contracts:
+
+- `patch` releases are to be used only for changes that do NOT modify contract bytecode (such as updating comments).
+- `minor` releases are to be used for changes that modify bytecode OR changes that expand the contract ABI provided that these changes do NOT break the existing interface.
+- `major` releases are to be used for changes that break the existing contract interface OR changes that modify the security model of a contract.
+
+Versioning for individual contracts:
+
+- A contract is only `X.Y.Z` on `develop` if it's been governance approved. If it's `X.Y.Z` before that, it must be on a branch. More on this below.
+- For contracts undergoing development, a `-beta.n` identifier must be appended to the version number.
+- For contracts in a release candidate state, an `-rc.n` identifier must be appended to the version number.
+- For contracts with feature-specific changes, a `+feature` identifier must be appended to the version number. See [ethereum-optimism/design-docs#8](https://github.com/ethereum-optimism/design-docs/pull/8) to learn more.
```

Comment:
Will there thus be `X.Y.Z-beta.n+feature-abc` tagged contracts?

Reply:
Sorry, `+feature` was meant to indicate `feature` as the variable name, maybe `+FEATURE-NAME` would be better? So in your example with a feature named `abc` it would be `X.Y.Z-beta.n+abc`

Reply:
Changed to `+feature-name` in 0f9f896 to match https://github.com/ethereum-optimism/design-docs/blob/main/smart-contract-feature-development.md